Add Heat Recovery Capability for Variable Refrigerant Flow Simulations

rongpengzhang opened the pull request:
## Pull request overview

The proposed new feature introduces an innovative model to simulate the energy performance of Variable Refrigerant Flow (VRF) systems with heat recovery (HR) configurations, which is capable of achieving heat recovery from cooling zones to heating zones and providing cooling and heating operation simultaneously. This feature is a continuation of the new VRF heat pump (HP) system model implemented in EnergyPlus V8.4.
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
  - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
  - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
  - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog
### Review Checklist

This will not be exhaustively relevant to every PR.
- [ ] Code style (parentheses padding, variable names)
- [ ] Functional code review (it has to work!)
- [ ] If defect, results of running current develop vs this branch should exhibit the fix
- [ ] CI status: all green or justified
- [ ] Performance: CI Linux results include performance check -- verify this
- [ ] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [ ] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [ ] Verify naming conventions and styles, memos and notes and defaults
  - [ ] If transition, add rules to spreadsheet
  - [ ] If transition, add transition source
  - [ ] If transition, update idfs
- [ ] If new idf included, locally check the err file and other outputs
- [ ] Documentation changes in place
- [ ] ExpandObjects changes??
- [ ] If output changes, including tabular output structure, add to output rules file for interfaces

rongpengzhang committed:
NFP-NewVRFHRModel

rongpengzhang committed:
NFP picture renaming

Pictures cannot be correctly displayed. Change picture name from
"xx.PNG"  to "xx.png"

rongpengzhang committed:
NFP picture renaming

rongpengzhang committed:
NFP picture update

NewVRFHeatRecoveryModel_Table2.PNG is updated to correct Mode 5 HP chart

rongpengzhang committed:
DD_NewVRFHRModel

rongpengzhang committed:
DD-Address comments in DD review meeting

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 2245154-NewVRFHRModel

rongpengzhang committed:
Idd and src

rongpengzhang committed:
Address build warnings

rongpengzhang committed:
Address build warnings

rongpengzhang committed:
Minor modifications

rongpengzhang committed:
Code clean up

rongpengzhang committed:
Example File

temp

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 2245154-NewVRFHRModel

rongpengzhang committed:
IO Reference, Parameter clean up

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 2245154-NewVRFHRModel

Myoldmopar commented:
@rongpengzhang Very first comments:
- I changed the title of the pull request so it is more digestible by users.  Please verify the name and change as you see fit.
- There are build warnings and they all appear to be due to ordered initialization issues.  This implies that in your class/struct definition, you have variables defined in the member initializer list in a different order than the variable is found defined in the class/struct itself.  If you can build on a Mac or Ubuntu, you should see these compiler warnings with a little more detail.  If you can't, you should be able to infer the issue from the CI results.  If you still can't, I can help resolve them.

I'm anxious to start reviewing this, so at some point I'll rigor over the code changes and you'll get 1 (one) zillion emails really quickly.  But it'll probably be tomorrow morning.

rongpengzhang commented:
@Myoldmopar, I have cleaned up the hh file to ensure that variables in the member initializer are in the same order with those in the class/struct member definition. I did the build on Windows, so let's see the CI results.

rongpengzhang committed:
hh file clean up

rongpengzhang committed:
Address build warnings

rongpengzhang commented:
@Myoldmopar VRF-HP and VRF-HR are two types of systems with very different system configurations and controls. That's why we develop a new model to simulate it, and add a new IDD object to collect system information. "Add Heat Recovery Capability for ..." sounds like we are adding an additional HR component or curves to the current VRF-HP model, but in fact it is much more than that. Do you think this name may confuse some users?

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

Myoldmopar commented on src/EnergyPlus/FluidProperties.cc:
```diff
@@ -4487,6 +4487,137 @@ namespace FluidProperties {
+		// na
+
+		// DERIVED TYPE DEFINITIONS:
+		// na
```
As an FYI, several of these comment sections aren't relevant and not needed.  I think they remain in the old Fortran template, but in general, especially for new functions, don't add them.  As another point of guidance, moving forward we'll want local variables defined as close as possible to where they are used/needed.  No need to define the locals at the top of the function like we used to do in Fortran.

This function doesn't appear to be a great candidate to become a member function of a class, but I'll be looking for that as I comment along here.

Myoldmopar commented on src/EnergyPlus/HVACVariableRefrigerantFlow.hh:
```diff
@@ -1049,7 +1088,200 @@ namespace HVACVariableRefrigerantFlow {
+	
+	void
+	VRFOU_TeTc(
+		int const VRFCond, // Index to VRF outdoor unit
```
And I just found them.   All of these functions.  Basically every function that has a first index of VRFCond.  These should be moved into member functions of the VRF class/struct.  So here's what I would love, and it's identical to what we were doing in the plant refactor:
- Take VRFCond index out as an argument in both the hh and cc files
- In the cc file, inside the function, each time you use `VRF( VRFCond ).` or whatever the variable name is, change it to `this->` (we can clean this up later, naysayers...)
- Move this function definition inside the class/struct above here in the hh file
- In the cc file, change the definition so it is: `VRFCondenserEquipment::VRFOU_TeTc`

If you did that for all of these functions, it would be a simple, but big step forward in a more structured VRF code base.  After that is complete, we can take a higher pass and decide if we want to change some ownership/structure to a more OO basis or not.

rongpengzhang replied:
@Myoldmopar It's a good idea to start the VRF OO design from these new functions. I will refactor them soon.

rongpengzhang committed:
Eng Ref

rongpengzhang committed:
Unit Test

rongpengzhang committed:
OO Refactor

Put the new added functions related with VRF-FluidTCtrl model as the
member functions of struct VRFCondenserEquipment or
VRFTerminalUnitEquipment.

Unit tests are updated correspondingly.

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 2245154-NewVRFHRModel

rongpengzhang commented:
@Myoldmopar I have put the new added functions related with VRF-FluidTCtrl model as the
member functions of struct VRFCondenserEquipment or VRFTerminalUnitEquipment. Unit tests are updated correspondingly. CI test result seems fine. Please let me know if you have other comments.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 16 days since this pull request was last updated.

Myoldmopar commented on src/EnergyPlus/FluidProperties.cc:
```diff
@@ -4487,6 +4487,137 @@ namespace FluidProperties {
+			}
+			RefrigIndex = RefrigNum;
+		}
+		auto const & refrig( RefrigData( RefrigNum ) );
```
@rongpengzhang how often is this called during a typical VRF calculation?  If it isn't too frequently then I don't see any problem with it.  If it is, we should take a real hard look at improving the performance of this function.  Maybe @mbadams5 could provide a quick comment on this function in case it is a performance-critical piece.

rongpengzhang replied:
@Myoldmopar,  the function "GetSupHeatTempRefrig" determines the refrigerant temperature at superheated region, given the pressure and enthalpy levels. It will be ideal if we know the physical equations to perform the calculations directly, but I don't have them, so I designed iterations to calculate it indirectly by using the existing function "GetSupHeatEnthalpyRefrig". I think you are suggesting to improve the iteration approach, so I have introduced SolveRegulaFalsi method. I have designed an unit test to assure it runs correctly.

Myoldmopar commented on src/EnergyPlus/HVACVariableRefrigerantFlow.hh:
```diff
@@ -573,46 +592,244 @@ namespace HVACVariableRefrigerantFlow {
 			IUCondensingTemp( 44.0 ),
 			IUEvapTempLow( 4.0 ),
-			IUEvapTempHigh( 13.0 ),
+			IUEvapTempHigh( 15.0 ),
```
@rraustad Does this change in a default value have any effect on the other VRF model?

rraustad replied:
This is a new variable which the other VRF model does not use.

Myoldmopar replied:
Thanks @rraustad; I figured by the name it was unrelated to your model, but good to verify it.

Myoldmopar committed:
Remove trailing whitepsace

Myoldmopar commented:
`goto` statements.... 😢

Myoldmopar commented:
Error file is real messy for the heat recovery VRF...I'm going to push up a cleanup commit that just trims out a bunch of whitespace from the code, but I'm not going to touch this again until the error file issues are resolved.

```
Program Version,EnergyPlus, Version 8.5.0-2cae9cb299, YMD=2016.07.08 08:11,IDD_Version 8.5.0
   ** Warning ** Version: in IDF="8.4" not the same as expected="8.5"
   ************* Beginning Zone Sizing Calculations
   ** Warning ** Calculated design heating load for zone=SPACE5-1 is zero.
   **   ~~~   ** Check Sizing:Zone and ZoneControl:Thermostat inputs.
   ** Warning ** SizeDXCoil Coil:Cooling:DX:VariableRefrigerantFlow:FluidTemperatureControl TU2 VRF DX COOLING COIL
   **   ~~~   ** ...Rated Total Cooling Capacity [W] will be limited by the maximum rated volume flow per rated total capacity ratio.
   **   ~~~   ** ...DX coil volume flow rate ( m3/s ) = 0.300425
   **   ~~~   ** ...Requested capacity ( W ) = 4969.057
   **   ~~~   ** ...Requested flow/capacity ratio ( m3/s/W ) = 6.045E-005
   **   ~~~   ** ...Maximum flow/capacity ratio ( m3/s/W ) = 6.041E-005
   **   ~~~   ** ...Adjusted capacity ( W ) = 4973.102
   ** Warning ** SizeDXCoil Coil:Cooling:DX:VariableRefrigerantFlow:FluidTemperatureControl TU4 VRF DX COOLING COIL
   **   ~~~   ** ...Rated Total Cooling Capacity [W] will be limited by the maximum rated volume flow per rated total capacity ratio.
   **   ~~~   ** ...DX coil volume flow rate ( m3/s ) = 0.333396
   **   ~~~   ** ...Requested capacity ( W ) = 5420.826
   **   ~~~   ** ...Requested flow/capacity ratio ( m3/s/W ) = 6.150E-005
   **   ~~~   ** ...Maximum flow/capacity ratio ( m3/s/W ) = 6.041E-005
   **   ~~~   ** ...Adjusted capacity ( W ) = 5518.896
   ** Warning ** SizeDXCoil Coil:Cooling:DX:VariableRefrigerantFlow:FluidTemperatureControl TU5 VRF DX COOLING COIL
   **   ~~~   ** ...Rated Total Cooling Capacity [W] will be limited by the maximum rated volume flow per rated total capacity ratio.
   **   ~~~   ** ...DX coil volume flow rate ( m3/s ) = 1.510404
   **   ~~~   ** ...Requested capacity ( W ) = 22391.173
   **   ~~~   ** ...Requested flow/capacity ratio ( m3/s/W ) = 6.745E-005
   **   ~~~   ** ...Maximum flow/capacity ratio ( m3/s/W ) = 6.041E-005
   **   ~~~   ** ...Adjusted capacity ( W ) = 25002.562
   ** Warning ** SizeDXCoil Coil:Heating:DX:VariableRefrigerantFlow:FluidTemperatureControl TU5 VRF DX HEATING COIL
   **   ~~~   ** ...Rated Total Heating Capacity [W] will be limited by the maximum rated volume flow per rated total capacity ratio.
   **   ~~~   ** ...DX coil volume flow rate ( m3/s ) = 1.510404
   **   ~~~   ** ...Requested capacity ( W ) = 25002.562
   **   ~~~   ** ...Requested flow/capacity ratio ( m3/s/W ) = 6.041E-005
   **   ~~~   ** ...Maximum flow/capacity ratio ( m3/s/W ) = 6.041E-005
   **   ~~~   ** ...Adjusted capacity ( W ) = 25002.562
   ************* Testing Individual Branch Integrity
   ************* All Branches passed integrity testing
   ************* Testing Individual Supply Air Path Integrity
   ************* All Supply Air Paths passed integrity testing
   ************* Testing Individual Return Air Path Integrity
   ************* All Return Air Paths passed integrity testing
   ************* No node connection errors were found.
   ************* Beginning Simulation
   ** Warning ** AirConditioner:VariableRefrigerantFlow "VRF HEAT PUMP".
   **   ~~~   ** ...InitVRF: VRF Heat Pump Min/Max Outdoor Temperature in Heat Recovery Mode Limits have been exceeded and VRF heat recovery is disabled.
   **   ~~~   ** ... Outdoor Dry-Bulb Temperature                       = 26.069
   **   ~~~   ** ... Heat Recovery Minimum Outdoor Dry-Bulb Temperature = -20.000
   **   ~~~   ** ... Heat Recovery Maximum Outdoor Dry-Bulb Temperature = 26.000
   **   ~~~   ** ... Check VRF Heat Pump Min/Max Outdoor Temperature in Heat Recovery Mode limits.
   **   ~~~   **  During Warmup, Environment=CHICAGO_IL_USA ANNUAL COOLING 1% DESIGN CONDITIONS DB/MCWB, at Simulation time=07/25 09:00 - 09:15
   **   ~~~   ** ...the system will operate in heat pump mode when applicable.
   ************* Simulation Error Summary *************
   *************
   ************* ===== Recurring Error Summary =====
   ************* The following recurring error messages occurred.
   *************
   *************  ** Warning ** AirConditioner:VariableRefrigerantFlow "VRF HEAT PUMP" -- Exceeded VRF Heat Recovery min/max outdoor temperature limit error continues...
   *************  **   ~~~   **   This error occurred 6020 total times;
   *************  **   ~~~   **   during Warmup 5160 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=31.5  Min=26.06975
   *************
   ************* EnergyPlus Warmup Error Summary. During Warmup: 0 Warning; 0 Severe Errors.
   ************* EnergyPlus Sizing Error Summary. During Sizing: 1 Warning; 0 Severe Errors.
   ************* EnergyPlus Completed Successfully-- 6026 Warning; 0 Severe Errors; Elapsed Time=00hr 00min 11.92sec
```

rongpengzhang commented:
@Myoldmopar, Let me explain these warning messages, which can be put into three categories:

(1)*\* Warning *\* Calculated design heating load for zone=SPACE5-1 is zero
This warning is reasonable. The internal zone (SPACE5-1) in the example file has large internal gains (10W/ft2, a typical level for common server rooms) and thus has no or little heating loads during winter time. This will lead to more simultaneous heating and cooling operations which can better show the characteristics of VRF-HR.

(2)*\* Warning *\* ...InitVRF: VRF Heat Pump Min/Max Outdoor Temperature in Heat Recovery Mode Limits have been exceeded and VRF heat recovery is disabled.
This warning is reasonable. The user specifies the range of outdoor air for normal VRF HR operations (-20C~26C in the example file). If OA temperature is out of this range, the system will not work in that mode. This rule applies to both the VRF-SystemCurve model and VRF-FluidTCtrl model. This warning message just notifies the user of this severe scenario. If the users believe it is fine to expand this range for their VRF-HR systems, they can change the range limits and these warnings will be gone.

(3) *\* Warning *\* SizeDXCoil Coil:Heating:DX:VariableRefrigerantFlow:FluidTemperatureControl TU5 VRF DX HEATING COIL
This warning is given in the autosizing of VRF indoor unit coils. There is a maximum flow/capacity ratio ( 6.041E-005 m3/s/W ) for all types of coils. If the autosized values go beyond this ratio, the coil flow & capacity will be further modified. In the VRF-FluidTCtrl coils, however, the coil capacity is related with the IU curves that the user provides. I will double check it.

rongpengzhang committed:
Address comments

1. Improve iteration design in GetSupHeatTempRefrig method
2. Clean up unnecessary warnings

rongpengzhang committed:
Merge remote-tracking branch 'origin/2245154-NewVRFHRModel' into 2245154-NewVRFHRModel

# Conflicts:
#	src/EnergyPlus/FluidProperties.cc

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 2245154-NewVRFHRModel

rongpengzhang committed:
Minor modification

rongpengzhang committed:
Address build warnings

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 2245154-NewVRFHRModel

Myoldmopar committed:
Merge develop; resolve conflict

Myoldmopar commented:
@rongpengzhang This is looking really good.  The error file has reduced, so thanks for that.  I merged in develop to get rid of a documentation-only merge conflict.  Once that comes back clean I'll merge this in (unless @mjwitte or @rraustad have any remaining issues).
